### PR TITLE
Preview pose estimate

### LINKED ISSUE #8: Create a restore point for to undo pose estimates
To avoid messing up the pose it would be helpful to revert changes done by sending initial poses. Either by keeping track of the poses or saving the current pose.

### PR DESCRIPTION
**Pull request by Eman7C7**
Instead of sending directly a pose estimate, the user must now confirm
with Enter (or undo with Esc)

Closes #8

### CODE REVIEW

File: src/transformation.rs [hunk 1]
```diff
@@ -17,19 +16,26 @@ pub fn transform_relative_pt(
...
         tf.rotation.z,
     ));
-    let isometry = Isometry3::from_parts(tra, rot);
-    let transformed_point = isometry.transform_point(&Point3::new(pt.0, pt.1, 0.));
-    (transformed_point.x, transformed_point.y)
+    let (_roll, _pitch, yaw) = rot.euler_angles();
+    (tf.translation.x, tf.translation.y, yaw)
+}
+
+pub fn minimal_to_ros(x: f64, y: f64, yaw: f64) -> rosrust_msg::geometry_msgs::Transform {
```
Comment (carzum) [inline src/transformation.rs:30]: not a huge fan of 'minimal'. it doesn't feel generic enough to justify it in this module. I thought about offering this with optional parameters for xyz and rpm ( or qx, qz..).

but this will be tricky[
](https://vidify.org/blog/rust-parameters/)

Comment (carzum) [reply src/transformation.rs:30]: https://vidify.org/blog/rust-parameters/

File: src/app.rs [hunk 2]
```diff
@@ -147,6 +145,21 @@ impl App {
         self.zoom -= self.zoom_factor;
     }
 
+    pub fn move_pose_estimate(&mut self, x: f64, y: f64, yaw: f64) {
+        self.pose_estimate.2 += yaw;
```
Comment (carzum) [inline src/app.rs:149]: minimal should be a struct. it would make the coding much clearer. This would also allow some methods for it

File: src/config.rs [hunk 3]
```diff
@@ -17,7 +17,7 @@ pub struct ListenerConfig {
 #[derive(Debug, Serialize, Deserialize)]
 pub struct TermvizConfig {
     pub fixed_frame: String,
-    pub camera_frame: String,
```
Comment (carzum) [inline src/config.rs:20]: keep camera frame, as we want it anyways at some point?

**Review by carzum — CHANGES_REQUESTED** [review]
I am exited to check it out on my machine, the minimal pose is a little confusing and causes some weird functions, I think should be converted to a struct before merge.

File: src/initial_pose.rs [hunk 4]
```diff
@@ -1,66 +1,34 @@
 use std::sync::Arc;
```
Comment (carzum) [inline src/initial_pose.rs:1]: unused now

**Comment by carzum** [comment]
I tested it a little on a stage simulation. setting the pose estimate on an idle robot seems intuitive enough, also didn't have have a problem with huge offsets, so we can skip the ghost laser. The step size is still not intuitive though, but that shouldn't be part of this pr.

I noticed one problem: I don't seem to be able to reset the ghost once the robot drives away from it. I would expect the ghost to go back to the robot once esc is hit

File: Cargo.toml [hunk 5]
```diff
@@ -8,7 +8,7 @@ exclude = ["assets/*", ".github"]
 
 
 [dependencies]
-nalgebra = "*"
+nalgebra = "0.29.0"
```
Comment (MaxiMaerz) [inline Cargo.toml:11]: why this version?

File: src/app.rs [hunk 6]
```diff
@@ -25,15 +25,12 @@ pub enum AppModes {
...
-    tf_listener: Arc<rustros_tf::TfListener>,
-    static_frame: String,
-    axis_length: f64,
-) -> Vec<Line> {
-    let tf = tf_listener
-        .lookup_transform(&static_frame, "base_link", rosrust::Time::new())
-        .unwrap()
-        .transform;
+// 1mm and slightly less than half a degree
+fn poses_are_close(p1: &(f64, f64, f64), p2: &(f64, f64, f64)) -> bool {
```
Comment (MaxiMaerz) [inline src/app.rs:29]: the name suggest poses as input but this accepts a tuple of three. Change to positions

File: src/app.rs [hunk 7]
```diff
@@ -299,13 +317,32 @@ impl App {
...
+                for line in get_frame_lines(&base_link_pose, self.axis_length) {
                     ctx.draw(&line);
                 }
+                if !poses_are_close(&self.initial_pose, &self.pose_estimate) {
+                    let pose_estimate_ros = transformation::minimal_to_ros(
+                        self.pose_estimate.0,
+                        self.pose_estimate.1,
+                        self.pose_estimate.2,
+                    );
+                    let ghost_footprint =
```
Comment (MaxiMaerz) [inline src/app.rs:330]: i never know but do we need the assignment here?

File: src/config.rs (hunk 3 above)
Comment (MaxiMaerz) [reply src/config.rs:20]: i would argue that we better get rid of unused params and add them if we need to

File: src/transformation.rs (hunk 1 above)
Comment (MaxiMaerz) [reply src/transformation.rs:30]: The bane of not having transformations.py
I also dislike the name, something like get_position_vector_from_pose or extract vector
However, (not part of this PR) but maybe it makes sense to create a QOL class to handle this in another crate?

File: src/app.rs [hunk 8]
```diff
@@ -78,6 +77,15 @@ impl App {
             config.marker_array_topics,
             config.map_topics,
         );
+        let base_link_pose = tf_listener
```
Comment (MaxiMaerz) [inline src/app.rs:80]: One Big Question:
Why did you add this to the app?
I think it would be nice to have the math and logic in the initial pose publisher (maybe rename it) and offer a draw method to the App.
So we can keep the logic separate from the frontend.

**Review by MaxiMaerz — CHANGES_REQUESTED** [review]
Super nice feature!
Besides the small questions i wonder if we should not move the logic from app to the pose handler.

File: Cargo.toml (hunk 5 above)
Comment (Eman7C7) [reply Cargo.toml:11]: supports conversion to euler angle directly (it's the most recent version atm)

File: src/app.rs (hunk 6 above)
Comment (Eman7C7) [reply src/app.rs:29]: it's a 2D pose: x, y, yaw. But your confusion proves Carsten's point: we need a struct

File: src/app.rs (hunk 7 above)
Comment (Eman7C7) [reply src/app.rs:330]: I guess not

File: src/config.rs (hunk 3 above)
Comment (Eman7C7) [reply src/config.rs:20]: exactly: I agree with maxi

File: src/transformation.rs (hunk 1 above)
Comment (Eman7C7) [reply src/transformation.rs:30]: we are working in 2D here: the minimal representation of a pose is x, y, angle. The only more "generic" thing is a 3D representation. Is anybody ok with a struct like "Pose2D" (or propose a similar name)?

**Comment by Eman7C7** [comment]
The ghost not moving when the robot moves is sort of intended: the ghost should NOT move when the robot moves, BUT since atm we don't have a separate mode for sending a pose it looks super ugly if the robot starts moving and a ghost stays... Will try to think about a solution

File: src/transformation.rs (hunk 1 above)
Comment (carzum) [reply src/transformation.rs:30]: Pose2D is fine with me. Nalgebra also has 2d "poses": https://docs.rs/nalgebra/0.10.1/nalgebra/struct.Isometry2.html

**Comment by Eman7C7** [comment]
I fixed everything except Maxi's comment about leaving the logic in app. That would require moving the publisher inside the app. Let me know if I should do it in this PR or in a new one

File: src/app.rs (hunk 8 above)
Comment (Eman7C7) [reply src/app.rs:77]: Wait do you mean specifically why do I lookup a tf here? If yes it's to cache the transformation, since we actually use it multiple times below

**Review by carzum — APPROVED** [review]
lgtm! I would argue to move as much logic from the app to the modules in a seperate pr, or just open an issue, it doesn't hurt much.

**Review by MaxiMaerz — APPROVED** [review]
Ok maybe you can make a greater or equal for the nalgebra version :) splititng the pose eestimate can be done in another PR

**Comment by carzum** [comment]
> Ok maybe you can make a greater or equal for the nalgebra version :) splititng the pose eestimate can be done in another PR

Cargo should do that automatically for  to some degree:
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html
